presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java

```diff
@@ -202,14 +211,39 @@ private static InternalNode chooseLeastBusyNode(List<InternalNode> candidateNode
 
             // choose the preferred node first as long as they're not busy
             if (preferredNodeCount.isPresent() && i < preferredNodeCount.getAsInt() && splitCount < maxSplitCount) {
```

Comment:
If I read this condition correctly, this will make HARD_AFFINITY to return split that is not cacheable. Can we add tests to verify that?